// @rafters/astro-meta/dist/og.js
// @rafters/astro-meta/og — Open Graph image generation
//
// Satori + resvg-js produce one PNG per (module, route) match at build:done.
// Both peers are optional; consumers that don't import this subpath pay
// nothing. v0.1 supports route filtering via OgModule.match; per-page input
// validation and cache-by-hash arrive in v0.2.
async function loadSatori() {
    try {
        return (await import("satori"));
    }
    catch {
        return null;
    }
}
async function loadResvg() {
    try {
        return (await import("@resvg/resvg-js"));
    }
    catch {
        return null;
    }
}
const DEFAULT_WIDTH = 1200;
const DEFAULT_HEIGHT = 630;
/** Render a single OG module to PNG bytes for the given context. */
export async function renderOg(module, ctx) {
    const [satoriMod, resvgMod] = await Promise.all([loadSatori(), loadResvg()]);
    if (!satoriMod || !resvgMod) {
        throw new Error("@rafters/astro-meta/og: satori and @resvg/resvg-js are required peer dependencies; install both to use the og subpath");
    }
    const element = module.template({ ctx });
    const fonts = module.fonts ?? [];
    const svg = await satoriMod.default(element, {
        width: module.width ?? DEFAULT_WIDTH,
        height: module.height ?? DEFAULT_HEIGHT,
        fonts,
    });
    const resvg = new resvgMod.Resvg(svg);
    return resvg.render().asPng();
}
/** Slug a route for use as a filename: "/" -> "index", "/about" -> "about", "/blog/post-1" -> "blog/post-1". */
export function ogSlugForRoute(route) {
    const trimmed = route.replace(/^\/+|\/+$/g, "");
    return trimmed.length === 0 ? "index" : trimmed;
}
