// @casoon/astro-site-files/dist/integration.js
import { mkdir, open, readFile, stat, writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { deduplicateEntries, resolveEntry } from './sitemap/compile.js';
import { renderSitemapIndex, renderSitemapXml } from './sitemap/render.js';
import { auditSitemap } from './sitemap/audit.js';
import { renderRssFeed } from './rss.js';
import { renderRobotsTxt } from './robots.js';
import { renderLlmsTxt } from './llms.js';
import { renderSecurityTxt } from './security.js';
import { renderHumansTxt } from './humans.js';
import { auditHumans, auditLlms, auditRobots, auditSecurity, filterIssues } from './audit.js';
const TODAY = new Date().toISOString().split('T')[0];
const PLUGIN = '@casoon/astro-site-files';
const BUILT_IN_SKIP = [
    /^\/sitemap(-index)?\.xml$/,
    /^\/robots\.txt$/,
    /^\/llms\.txt$/,
    /^\/rss\.xml$/,
    /^\/feed\.xml$/,
    /\/404\b/,
    /\/500\b/,
    /^\/_/,
    /\/api\//,
    /\/landing\//,
    /\/drafts\//,
];
// ── Helpers ───────────────────────────────────────────────────────────────────
function buildSiteWithBase(siteUrl, base) {
    const normalizedBase = (!base || base === '/') ? '' : '/' + base.replace(/^\/|\/$/g, '');
    return (siteUrl?.replace(/\/$/, '') ?? '') + normalizedBase;
}
function shouldSkip(urlPath, userExcludes, userFilter, fullUrl) {
    if (BUILT_IN_SKIP.some(p => p.test(urlPath)))
        return true;
    if (userExcludes.some(p => typeof p === 'string'
        ? urlPath === p || urlPath.startsWith(p)
        : p.test(urlPath)))
        return true;
    if (userFilter && fullUrl && !userFilter(fullUrl))
        return true;
    return false;
}
function buildI18nLinks(entries, i18n, siteUrl) {
    const { defaultLocale, locales } = i18n;
    const localeKeys = Object.keys(locales);
    const base = siteUrl.replace(/\/$/, '');
    function getCanonicalInfo(fullUrl) {
        const path = fullUrl.startsWith(base) ? fullUrl.slice(base.length) : fullUrl;
        const normalized = path.startsWith('/') ? path : `/${path}`;
        for (const locale of localeKeys) {
            const prefix = `/${locale}`;
            if (normalized === prefix || normalized.startsWith(`${prefix}/`)) {
                return { canonical: normalized.slice(prefix.length) || '/', locale };
            }
        }
        return { canonical: normalized, locale: defaultLocale };
    }
    const groups = new Map();
    for (const entry of entries) {
        const { canonical, locale } = getCanonicalInfo(entry.loc);
        if (!groups.has(canonical))
            groups.set(canonical, []);
        groups.get(canonical).push({ locale, entry });
    }
    return entries.map(entry => {
        const { canonical } = getCanonicalInfo(entry.loc);
        const group = groups.get(canonical) ?? [];
        if (group.length <= 1)
            return entry;
        const links = group.map(({ locale, entry: e }) => ({
            hreflang: locales[locale] ?? locale,
            href: e.loc,
        }));
        return { ...entry, links };
    });
}
async function fileInfo(outDir, urlPath) {
    const filePath = urlPath === '/' || urlPath.endsWith('/')
        ? join(outDir, urlPath, 'index.html')
        : join(outDir, urlPath);
    const s = await stat(filePath).catch(() => null);
    if (!s)
        return { isRedirect: false };
    const lastmod = s.mtime.toISOString().split('T')[0];
    // Read the first 512 bytes to detect meta-refresh redirect pages
    const fh = await open(filePath, 'r').catch(() => null);
    if (!fh)
        return { lastmod, isRedirect: false };
    try {
        const buf = Buffer.alloc(512);
        const { bytesRead } = await fh.read(buf, 0, 512, 0);
        const head = buf.toString('utf8', 0, bytesRead).toLowerCase();
        const isRedirect = head.includes('<meta http-equiv="refresh"');
        return { lastmod, isRedirect };
    }
    finally {
        await fh.close();
    }
}
async function readHtmlMetadata(filePath) {
    const content = await readFile(filePath, 'utf-8').catch(() => '');
    if (!content)
        return {};
    // Extract JSON-LD script and search for custom changefreq/priority data-attributes
    const jsonLdMatch = /<script\s+[^>]*type=["']application\/ld\+json["']([^>]*?)>/i.exec(content);
    if (!jsonLdMatch)
        return {};
    const attrs = jsonLdMatch[1];
    const changefreqMatch = /data-sitemap-changefreq=["'](.*?)["']/i.exec(attrs);
    const priorityMatch = /data-sitemap-priority=["'](.*?)["']/i.exec(attrs);
    const res = {};
    if (changefreqMatch)
        res.changefreq = changefreqMatch[1];
    if (priorityMatch) {
        const val = parseFloat(priorityMatch[1]);
        if (!isNaN(val))
            res.priority = val;
    }
    return res;
}
// ── Integration ───────────────────────────────────────────────────────────────
export default function siteFiles(options = {}) {
    let astroConfig;
    let fallbackPathnames = [];
    return {
        name: PLUGIN,
        hooks: {
            'astro:config:setup'({ config }) {
                astroConfig = config;
            },
            'astro:routes:resolved'({ routes }) {
                if (options.sitemap === false)
                    return;
                fallbackPathnames = routes
                    .filter(r => r.type === 'page' && r.fallbackRoutes?.length > 0)
                    .flatMap(r => r.fallbackRoutes)
                    .filter(fr => fr.pathname != null)
                    .map(fr => fr.pathname);
            },
            async 'astro:build:done'({ pages, dir, logger, }) {
                const outDir = fileURLToPath(dir);
                await writeRobots(outDir, options, astroConfig, logger);
                await writeLlms(outDir, options, logger);
                await writeSecurity(outDir, options, logger);
                await writeHumans(outDir, options, logger);
                if (options.sitemap !== false) {
                    await writeSitemap(outDir, options, astroConfig, pages, fallbackPathnames, logger);
                }
            },
        },
    };
}
// ── File writers ──────────────────────────────────────────────────────────────
async function writeRobots(outDir, options, astroConfig, logger) {
    if (options.robots === false)
        return;
    const robotsOpts = typeof options.robots === 'object' ? options.robots : {};
    const siteUrl = astroConfig?.site ? buildSiteWithBase(String(astroConfig.site), astroConfig.base) : undefined;
    await writeFile(join(outDir, 'robots.txt'), renderRobotsTxt(robotsOpts, siteUrl), 'utf-8');
    logger.info('robots.txt generated');
    for (const issue of filterIssues(auditRobots(robotsOpts), options.audit)) {
        logger[issue.level](`[${issue.rule}] ${issue.message} — ${issue.help}`);
    }
}
async function writeLlms(outDir, options, logger) {
    if (!options.llms)
        return;
    if (options.llms === true) {
        logger.warn('llms: requires a title — provide an object with { title } to generate llms.txt');
        return;
    }
    const llmsOpts = options.llms;
    const sourceSections = [];
    for (const source of llmsOpts.sources ?? []) {
        sourceSections.push(await source());
    }
    const resolvedOpts = sourceSections.length
        ? { ...llmsOpts, sections: [...(llmsOpts.sections ?? []), ...sourceSections] }
        : llmsOpts;
    await writeFile(join(outDir, 'llms.txt'), renderLlmsTxt(resolvedOpts), 'utf-8');
    logger.info('llms.txt generated');
    for (const issue of filterIssues(auditLlms(resolvedOpts), options.audit)) {
        logger[issue.level](`[${issue.rule}] ${issue.message} — ${issue.help}`);
    }
}
async function writeSecurity(outDir, options, logger) {
    if (!options.security)
        return;
    if (options.security === true || !options.security.contact) {
        logger.warn('security: requires a contact field (RFC 9116) — skipping security.txt');
        return;
    }
    const wellKnownDir = join(outDir, '.well-known');
    await mkdir(wellKnownDir, { recursive: true });
    await writeFile(join(wellKnownDir, 'security.txt'), renderSecurityTxt(options.security), 'utf-8');
    logger.info('.well-known/security.txt generated');
    for (const issue of filterIssues(auditSecurity(options.security), options.audit)) {
        logger[issue.level](`[${issue.rule}] ${issue.message} — ${issue.help}`);
    }
}
async function writeHumans(outDir, options, logger) {
    if (!options.humans)
        return;
    const humansOpts = typeof options.humans === 'object' ? options.humans : {};
    await writeFile(join(outDir, 'humans.txt'), renderHumansTxt(humansOpts), 'utf-8');
    logger.info('humans.txt generated');
    for (const issue of filterIssues(auditHumans(humansOpts), options.audit)) {
        logger[issue.level](`[${issue.rule}] ${issue.message} — ${issue.help}`);
    }
}
async function writeSitemap(outDir, options, astroConfig, pages, fallbackPathnames, logger) {
    const sitemapOpts = typeof options.sitemap === 'object' ? options.sitemap : {};
    const siteUrl = (sitemapOpts.siteUrl ?? (astroConfig?.site ? String(astroConfig.site) : undefined))?.replace(/\/$/, '');
    const effectiveSiteUrl = buildSiteWithBase(siteUrl, astroConfig?.base);
    // ── Collect static pages ──────────────────────────────────────────────────
    const allPathnames = new Set([
        ...pages.filter(p => p?.pathname != null).map(p => p.pathname),
        ...fallbackPathnames,
    ]);
    const staticEntries = [];
    for (const raw of allPathnames) {
        let urlPath = raw === '' ? '/' : raw.startsWith('/') ? raw : `/${raw}`;
        if ((astroConfig?.build?.format === 'file' || astroConfig?.build?.format === 'preserve') &&
            urlPath !== '/' &&
            !urlPath.endsWith('/')) {
            urlPath = `${urlPath}.html`;
        }
        const fullUrl = effectiveSiteUrl ? `${effectiveSiteUrl}${urlPath}` : urlPath;
        if (shouldSkip(urlPath, sitemapOpts.exclude ?? [], sitemapOpts.filter, fullUrl))
            continue;
        const { lastmod, isRedirect } = await fileInfo(outDir, urlPath);
        if (isRedirect)
            continue;
        // Extract optional changefreq/priority set dynamically by the pages
        const filePath = urlPath === '/' || urlPath.endsWith('/')
            ? join(outDir, urlPath, 'index.html')
            : join(outDir, urlPath);
        const htmlMeta = await readHtmlMetadata(filePath);
        staticEntries.push({
            loc: fullUrl,
            lastmod,
            ...(htmlMeta.changefreq ? { changefreq: htmlMeta.changefreq } : {}),
            ...(htmlMeta.priority !== undefined ? { priority: htmlMeta.priority } : {}),
        });
    }
    // ── Collect dynamic sources ───────────────────────────────────────────────
    const sourceEntries = [];
    for (const source of sitemapOpts.sources ?? []) {
        sourceEntries.push(...await source());
    }
    // ── Resolve ───────────────────────────────────────────────────────────────
    const allRaw = [...staticEntries, ...sourceEntries];
    const resolved = allRaw.map(e => resolveEntry(e, sitemapOpts, effectiveSiteUrl));
    // ── Audit before dedup so duplicate URLs are detectable ───────────────────
    const issues = auditSitemap(resolved, sitemapOpts, siteUrl);
    // ── Deduplicate, serialize ────────────────────────────────────────────────
    let entries = deduplicateEntries(resolved);
    if (sitemapOpts.serialize) {
        const serialized = [];
        for (const entry of entries) {
            const result = await sitemapOpts.serialize(entry);
            if (result !== undefined)
                serialized.push(result);
        }
        entries = deduplicateEntries(serialized);
    }
    // ── i18n hreflang ─────────────────────────────────────────────────────────
    if (sitemapOpts.i18n && effectiveSiteUrl) {
        entries = buildI18nLinks(entries, sitemapOpts.i18n, effectiveSiteUrl);
    }
    // ── Log audit issues ──────────────────────────────────────────────────────
    for (const issue of filterIssues(issues, options.audit)) {
        logger[issue.level]?.(`[${issue.rule}] ${issue.message} — ${issue.help}`);
    }
    // ── Write output ──────────────────────────────────────────────────────────
    await mkdir(outDir, { recursive: true });
    const maxUrls = sitemapOpts.output?.maxUrls ?? 50_000;
    const filename = sitemapOpts.output?.filename ?? 'sitemap.xml';
    const useIndex = sitemapOpts.output?.mode === 'index' || entries.length > maxUrls;
    if (!useIndex) {
        const xml = renderSitemapXml(entries, `Generated by ${PLUGIN}`);
        await writeFile(join(outDir, filename), xml, 'utf-8');
        logger.info(`sitemap.xml generated (${entries.length} URLs)`);
    }
    else {
        const chunks = [];
        for (let i = 0; i < entries.length; i += maxUrls) {
            chunks.push(entries.slice(i, i + maxUrls));
        }
        const indexEntries = [];
        for (let i = 0; i < chunks.length; i++) {
            const chunkFile = `sitemap-${i + 1}.xml`;
            const xml = renderSitemapXml(chunks[i], `Part ${i + 1}/${chunks.length} — Generated by ${PLUGIN}`);
            await writeFile(join(outDir, chunkFile), xml, 'utf-8');
            indexEntries.push({
                loc: effectiveSiteUrl ? `${effectiveSiteUrl}/${chunkFile}` : `/${chunkFile}`,
                lastmod: TODAY,
            });
        }
        const indexXml = renderSitemapIndex(indexEntries);
        await writeFile(join(outDir, 'sitemap-index.xml'), indexXml, 'utf-8');
        logger.info(`sitemap-index.xml generated (${chunks.length} parts, ${entries.length} total URLs)`);
    }
    if (sitemapOpts.rss) {
        const rss = sitemapOpts.rss;
        const rssFilename = rss.filename ?? 'rss.xml';
        const rssSiteUrl = effectiveSiteUrl ?? siteUrl ?? '';
        const rssItems = await rss.getItems(rssSiteUrl);
        const rssXml = renderRssFeed({
            title: rss.title,
            description: rss.description,
            feedUrl: rss.feedUrl ?? `${rssSiteUrl}/${rssFilename}`,
            language: rss.language,
            copyright: rss.copyright,
            managingEditor: rss.managingEditor,
            feedCustomData: rss.feedCustomData,
            xmlns: rss.xmlns,
        }, rssSiteUrl, rssItems);
        await writeFile(join(outDir, rssFilename), rssXml, 'utf-8');
        logger.info(`${rssFilename} generated (${rssItems.length} items)`);
    }
}
